Implemented basic authentication and authorization capabilities

Whitomtit opened the pull request:
Initial implementation of a library module for our microservices, that contains all features from TekClinic/backlog#23

Whitomtit committed:
Implemented basic authentication and authorization capabilities, close TekClinic/backlog#23

vMaroon commented on ms.go:
```diff
@@ -0,0 +1,140 @@
+
+// GetRequiredEnv retrieves the value of the environment variable named by the key.
+// If the variable is not present in the environment an error is returned.
+func GetRequiredEnv(key string) (string, error) {
```
Should this function really be exported? The same applies to `GetOptionalEnv` and `FetchServiceParameters`. Recall that capitalization means exportation in go.

Whitomtit replied:
The reason these functions are exported because they are reused across different microservices.
For example, if we define additional configuration parameters in environment variable like database host/driver, timeouts etc, we have to implement `GetRequiredEnv`/`GetOptionalEnv` ourselves again.

In the case of `FetchServiceParameters`, if some microservice wants to talk with another microservice, it has to know its host and port, that should be configurable.

vMaroon replied:
It might be time to use etcd for service registration and discovery. We can keep this for now, but this will need to be done this semester at some point.

vMaroon commented on ms.go:
```diff
@@ -0,0 +1,140 @@
+func GetRequiredEnv(key string) (string, error) {
+	value, set := os.LookupEnv(key)
+	if !set {
+		return "", errors.New(key + " environment variable is missing")
```
Use [fmt.Errorf](https://cs.opensource.google/go/go/+/go1.22.0:src/fmt/errors.go;l=22) for creating errors (in this case a formatted error).

vMaroon commented on ms.go:
```diff
@@ -0,0 +1,140 @@
+		return false
+	}
+
+	for _, r := range roles {
```
Optional: you can create a [Set](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/sets#Set) for roles and use `.Has`.

Whitomtit replied:
@vMaroon is it a good idea to have the whole `apimachinery` as dependency only for Set?
Maybe `slices.Contains` can be used instead?

vMaroon replied:
Modules size is not a concern here, we're not running on an edge device. 

While this code is readable and not using sets.Set is not a deal breaker, the intention was to introduce you to this package in order to utilize Set when possible, for _usually_ less code lines and more readability.

vMaroon commented on ms.go:
```diff
@@ -0,0 +1,140 @@
+// Claims contains claims from a token and allows roles-checking for authorization.
+type Claims struct {
+	PreferredUsername string                         `json:"preferred_username"`
+	ResourceAccess    map[string]map[string][]string `json:"resource_access"`
```
Nested maps should be avoided when possible, is it possible? (E.g, by defining a supporting struct)

Whitomtit replied:
Claims are basically a JSON and it's its structure.
Actually user of the lib doesn't need the direct access to `ResourceAccess` and `RealmAccess`.
Instead, we can store just a set of roles (strings) that are defined in the same way as a parameter of `HasRole`.
Will it be a more suitable solution?

vMaroon replied:
Let's work under the principle of least privilege wherever possible, so wherever it is possible and makes sense to hide information from "user" of something, we'll do it.

What you wrote sounds reasonable.

BTW: is this aligned with how Keycloak's authorization works?

Whitomtit replied:
It is aligned with how Keycloak's authorization works.
There is no specific claim in JWT token that is responsible for authorization in RFC. I will extract the code that is responsible for translation claims into roles in a separate function, so it will be easier to expand it for another provider.

vMaroon commented on ms.go:
```diff
@@ -0,0 +1,140 @@
+		return nil, err
+	}
+
+	port := GetOptionalEnv(fmt.Sprintf("MS_%s_PORT", strings.ToUpper(serviceName)), "9090")
```
Let's avoid setting such constants in here, move to const in top of file.

vMaroon commented on ms.go:
```diff
@@ -0,0 +1,140 @@
+}
+
+// BaseServiceServer contains a port on which the server listens for GRPC requests and provides JWT token verification.
+type BaseServiceServer struct {
```
It feels like `BaseServiceServer` and `Claims` can be interfaced, I'm not sure though, depends on how you're going to use them.

Whitomtit replied:
@vMaroon Aren't interface suitable only for method specification? 
Use case of `BaseServiceServer`, `Claims` and `GetRequiredEnv`:
We want to implement some MS that has additional parameter which will be provided via env variable `EXTRA_FIELD`.

Firstly we define a struct for our MS that promotes:
```go
type MyServer struct {
	*BaseServiceServer
	additionalField string
}
```

It will be initialized in the following way:
```go
var s MyServer
s.BaseServiceServer, _ = CreateBaseServiceServer()
s.additionalField, _ = GetRequiredEnv("EXTRA_FIELD")
```

Implementation of `someGRPCFunc` that requires authentication and authorization:
```go
func (server MyServer) someGRPCFunc(ctx context.Context, token string) {
	// check authentication
	claims, err := server.VerifyToken(ctx, token)
	if err != nil {
		// some error
	}

	// check authorization
	authorized := claims.HasRole("doctor")
	if !authorized {
		// some error
	}
}
```

vMaroon replied:
I don't see how the example you provided does not work if `BaseServiceServer` was an interface.
You would:
1. Change the name of the implemented `BaseServiceServer` struct (e.g., to `baseServiceServer`)
2. Declare `BaseServiceServer` interface with the exposed functions
3. Implement `func NewBaseServiceServer(params if needed) BaseServiceServer` through initializing and returning a `baseServiceServer`

This way, the member methods would be more easily readable, and the inside logic of `baseServiceServer` would be protected.

The change to `MyServer` would be that you'd initialize its `BaseServiceServer` member through the new func.

vMaroon reviewed (changes requested):
Good work overall, some minor comments left.

vMaroon commented:
BTW, im reviewing you under open source practices im familiar with, and the personal preferences and styling im adopting from an extremely skilled professional who made tons of significant open source contributions to high end projects, who reviews my work.

Whitomtit committed:
Code enhancements to meet PR requirements
Code is organized in separate files.
BaseServiceServer and Claims transformed to interfaces.
Set-based roles. Added another level of abstraction to Claims.

vMaroon commented on server.go:
```diff
@@ -0,0 +1,145 @@
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+const defaultGRPCPort = "9090"
+const envAuthIssuerURL = "AUTH_ISSUER"
+const envGRPCPort = "GRPC_PORT"
+
+const authAuditor = "account"
+
+const rolesSeparator = "."
+const rolesKey = "roles"
```
```suggestion
const (
	defaultGRPCPort = "9090"
	envAuthIssuerURL = "AUTH_ISSUER"
	envGRPCPort = "GRPC_PORT"
	
	authAuditor = "account"
	
	rolesSeparator = "."
	rolesKey = "roles"
)
```

vMaroon reviewed (changes requested):
Looks good, one last comment

Whitomtit committed:
Grouped constants

Co-authored-by: Maroon Ayoub <maroon.ayoub@ibm.com>
